examples/esp32-with-cred/esp32-server-with-cred.rs

```diff
@@ -0,0 +1,246 @@
+use log::*;
+use thiserror::Error;
+
+use esp_idf_svc::eventloop::EspSystemEventLoop;
+use micro_rdk::{
+    common::{app_client::AppClientConfig, entry::RobotRepresentation},
+    esp32::{certificate::WebRtcCertificate, entry::serve_web, tls::Esp32TlsServerConfig},
+};
+
+#[cfg(feature = "qemu")]
+use {
+    micro_rdk::{
+        common::{
+            board::FakeBoard,
+            robot::{LocalRobot, ResourceMap, ResourceType},
+        },
+        proto::common::v1::ResourceName,
+    },
+    std::{
+        collections::HashMap,
+        net::Ipv4Addr,
+        sync::{Arc, Mutex},
+    },
+};
+
+#[cfg(not(feature = "qemu"))]
+use {
+    embedded_svc::wifi::{
+        AuthMethod, ClientConfiguration as WifiClientConfiguration,
+        Configuration as WifiConfiguration,
+    },
+    esp_idf_hal::{peripheral::Peripheral, prelude::Peripherals},
+    esp_idf_svc::nvs::{EspDefaultNvs, EspDefaultNvsPartition, EspNvs},
+    esp_idf_svc::wifi::{BlockingWifi, EspWifi},
+    esp_idf_sys as _,
+    esp_idf_sys::{esp_wifi_set_ps, EspError},
+    micro_rdk::common::registry::ComponentRegistry,
+};
+
+#[derive(Debug, Error)]
+pub enum ServerError {
+    #[error("Error fetching NVS key: {0}")]
+    NVSKeyError(String),
+    #[error("{0}")]
+    EspError(EspError),
+    #[error("Error obtaining peripherals")]
+    PeripheralsError,
+}
+
+impl From<EspError> for ServerError {
+    fn from(value: EspError) -> ServerError {
+        ServerError::EspError(value)
+    }
+}
+
+const VIAM_NVS_NAMESPACE: &str = "VIAM_NS";
+
+#[cfg(not(feature = "qemu"))]
+fn get_str_from_nvs(viam_nvs: &EspDefaultNvs, key: &str) -> Result<String, ServerError> {
+    let mut buffer_ref = [0_u8; 4000];
+    Ok(viam_nvs
+        .get_str(key, &mut buffer_ref)?
+        .ok_or(ServerError::NVSKeyError(key.to_string()))?
+        .trim_matches(char::from(0))
+        .to_string())
+}
+
+#[cfg(not(feature = "qemu"))]
+fn get_blob_from_nvs(viam_nvs: &EspDefaultNvs, key: &str) -> Result<Vec<u8>, ServerError> {
+    let mut buffer_ref = [0_u8; 4000];
+    Ok(viam_nvs
+        .get_blob(key, &mut buffer_ref)?
+        .ok_or(ServerError::NVSKeyError(key.to_string()))?
+        .to_vec())
+}
+
+#[cfg(not(feature = "qemu"))]
+struct NvsStaticVars {
+    wifi_ssid: String,
+    wifi_pwd: String,
+    robot_secret: String,
+    robot_id: String,
+    robot_dtls_cert: Vec<u8>,
+    robot_dtls_key_pair: Vec<u8>,
+    robot_dtls_cert_fp: String,
+    robot_srv_pem_chain: Vec<u8>,
+    robot_srv_pem_ca: Vec<u8>,
+    robot_srv_der_key: Vec<u8>,
+}
+
+impl NvsStaticVars {
+    fn new() -> Result<NvsStaticVars, ServerError> {
+        let nvs = EspDefaultNvsPartition::take()?;
+        info!("get namespace...");
+        let viam_nvs = EspNvs::new(nvs.clone(), VIAM_NVS_NAMESPACE, true)?;
+        info!("loading creds...");
+        let res = Ok(NvsStaticVars {
```

Comment:
(super nit, totally fine to ignore) don't need to assign this to `res` and then return `res`, you can just construct the return value